Add vLLM docs for Aurora (#727)
argonne-lcf/user-guides: docs/aurora/data-science/inference/vllm.md

# Inference with vLLM

vLLM is an open-source library designed to optimize the inference and serving. Originally developed at UC Berkeley's Sky Computing Lab, it has evolved into a community-driven project. The library is built around the innovative PagedAttention algorithm, which significantly improves memory management by reducing waste in Key-Value (KV) cache memory.

## Install vLLM 

> Install vllm using pre-built wheels. 

Reviewer: ```suggestion
Install vLLM using pre-built wheels. 
```

> ```bash

Reviewer: ```suggestion
```bash linenums="1"
```

conda create --name vllm python=3.10 -y
conda activate vllm

module unload oneapi/eng-compiler/2024.07.30.002
module use /opt/aurora/24.180.3/spack/unified/0.8.0/install/modulefiles/oneapi/2024.07.30.002
module use /soft/preview/pe/24.347.0-RC2/modulefiles
module add oneapi/release

pip install /flare/datascience/sraskar/vllm-install/wheels/*
pip install /flare/datascience/sraskar/vllm-install/vllm-0.6.6.post2.dev28+g5dbf8545.d20250129.xpu-py3-none-any.whl
```

## Access Model Weights

Model weights for commonly used open-weight models are downloaded and available in the following directory on Aurora:
```bash
/flare/datascience/model-weights/hub
```
To ensure your workflows utilize the preloaded model weights and datasets, update the following environment variables in your session. Some models hosted on Hugging Face may be gated, requiring additional authentication. To access these gated models, you will need a [Hugging Face authentication token](https://huggingface.co/docs/hub/en/security-tokens).
```bash
export HF_HOME="/flare/datascience/model-weights/hub"
export HF_DATASETS_CACHE="/flare/datascience/model-weights/hub"
export HF_TOKEN="YOUR_HF_TOKEN"
```

> ## Common Configuration Recommendations 

Reviewer: can you give more details/examples of typical models and model sizes (in parameter counts) that typically fit in a single tile?

For small models that fit within a single tile's memory (64 GB), no additional configuration is required to serve the model. Simply set `TP=1` (Tensor Parallelism). This configuration ensures the model is run on a single tile without the need for distributed setup.

> To utilize multiple tiles for larger models (`TP>1`), a more advanced setup is necessary. This involves configuring a Ray cluster and setting the `ZE_FLAT_DEVICE_HIERARCHY` environment variable.
> ```bash

Reviewer: ```suggestion
To utilize multiple tiles for larger models (`TP>1`), a more advanced setup is necessary. This involves configuring a Ray cluster and setting the `ZE_FLAT_DEVICE_HIERARCHY` environment variable.
```bash linenums="1"
```

export ZE_FLAT_DEVICE_HIERARCHY=FLAT

export VLLM_HOST_IP=$(getent hosts $(hostname).hsn.cm.aurora.alcf.anl.gov | awk '{ print $1 }' | tr ' ' '\n' | sort | head -n 1)
export tiles=12
ray --logging-level debug start --head --verbose --node-ip-address=$VLLM_HOST_IP --port=6379 --num-cpus=64 --num-gpus=$tiles&

> ```
> !!! info "DLPack notes on Aurora"
> 	- The number of GPU devices visible on each node depends on the `ZE_FLAT_DEVICE_HIERARCHY` environment variable. When ZE_FLAT_DEVICE_HIERARCHY=FLAT, the system exposes 12 devices (tiles).
>     - The --num-gpus parameter in the Ray cluster setup command should match the number of tiles you intend to use on each node.

Reviewer: ```suggestion
```

!!! info "DLPack notes on Aurora"
	
    - The number of GPU devices visible on each node depends on the `ZE_FLAT_DEVICE_HIERARCHY` environment variable. When `ZE_FLAT_DEVICE_HIERARCHY=FLAT`, the system exposes 12 devices (tiles).
    - The `--num-gpus` parameter in the Ray cluster setup command should match the number of tiles you intend to use on each node.

```

> ## Serve Small Models 
> 
> #### Using Single Tile

Reviewer: ```suggestion
## Serving Small Models 

#### Using a Single Tile
```

> Following serves `meta-llama/Llama-2-7b-chat-hf` on single tile of single node. 
> ```bash
> vllm serve meta-llama/Llama-2-7b-chat-hf --port 8000 --device xpu --dtype float16
> ```

Reviewer: ```suggestion
The following command serves `meta-llama/Llama-2-7b-chat-hf` on a single tile of a single node:
```bash linenums="1"
vllm serve meta-llama/Llama-2-7b-chat-hf --port 8000 --device xpu --dtype float16
```
```

#### Using Multiple Tiles

> Following sets up ray cluster and servers `meta-llama/Llama-2-7b-chat-hf` on 8 tiles on single node. 
> ```bash
> export VLLM_HOST_IP=$(getent hosts $(hostname).hsn.cm.aurora.alcf.anl.gov | awk '{ print $1 }' | tr ' ' '\n' | sort | head -n 1)
> ray --logging-level debug start --head --verbose --node-ip-address=$VLLM_HOST_IP --port=6379 --num-cpus=64 --num-gpus=8&
> vllm serve meta-llama/Llama-2-7b-chat-hf --port 8000 --tensor-parallel-size 8 --device xpu --dtype float16 --trust-remote-code
> ```

Reviewer: ```suggestion
The following commands set up the Ray cluster and serve `meta-llama/Llama-2-7b-chat-hf` on 8 tiles on single node. 
```bash linenums="1"
export VLLM_HOST_IP=$(getent hosts $(hostname).hsn.cm.aurora.alcf.anl.gov | awk '{ print $1 }' | tr ' ' '\n' | sort | head -n 1)
ray --logging-level debug start --head --verbose --node-ip-address=$VLLM_HOST_IP --port=6379 --num-cpus=64 --num-gpus=8&
vllm serve meta-llama/Llama-2-7b-chat-hf --port 8000 --tensor-parallel-size 8 --device xpu --dtype float16 --trust-remote-code
```
```

> ## Serve Medium Models 
> 
> #### Using single Node

Reviewer: ```suggestion
## Serving Medium Models 

#### Using a Single Node
```

Reviewer 2: done

> Following sets up ray cluster and servers `meta-llama/Llama-3.3-70B-Instruct` on 8 tiles on single node. 
> ```bash

Reviewer: ```suggestion
The following commands set up the Ray cluster and serve `meta-llama/Llama-3.3-70B-Instruct` on 8 tiles on a single node:
```bash linenums="1"
```

export VLLM_HOST_IP=$(getent hosts $(hostname).hsn.cm.aurora.alcf.anl.gov | awk '{ print $1 }' | tr ' ' '\n' | sort | head -n 1)
ray --logging-level debug start --head --verbose --node-ip-address=$VLLM_HOST_IP --port=6379 --num-cpus=64 --num-gpus=8&
vllm serve meta-llama/Llama-3.3-70B-Instruct --port 8000 --tensor-parallel-size 8 --device xpu --dtype float16 --trust-remote-code
```

#### Using Multiple Nodes

> Use [setup_ray_cluster.sh](./setup_ray_cluster.sh) script to setup ray cluster across nodes. 

Reviewer: isnt this redundant with the next section?

Reviewer 2: Agreed, removed it.

Following serves `meta-llama/Llama-3.3-70B-Instruct` using 2 nodes with `TP=8` and `PP=2` 

```bash
vllm serve meta-llama/Llama-3.3-70B-Instruct --port 8000 --tensor-parallel-size 8 --pipeline-parallel-size 2 --device xpu --dtype float16 --trust-remote-code
```

## Serve Large Models 

> Use [setup_ray_cluster.sh](./setup_ray_cluster.sh) script to setup ray cluster across nodes. 

Reviewer: ```suggestion
Use [setup_ray_cluster.sh](./setup_ray_cluster.sh) script to setup a Ray cluster across nodes. 
```

And change the link to a path in `../../../GettingStarted/` after moving the script?

Reviewer 2: @sraskar Can you also provide the run_ray_cluster.sh script. This should run ray head and ray workers across nodes. We also would need to specify how to fetch nodes through qsub. I can work on it if you provide the run script

```bash linenums="1"
export ZE_FLAT_DEVICE_HIERARCHY=FLAT

export VLLM_HOST_IP=$(getent hosts $(hostname).hsn.cm.aurora.alcf.anl.gov | awk '{ print $1 }' | tr ' ' '\n' | sort | head -n 1)
export tiles=12
ray --logging-level debug start --head --verbose --node-ip-address=$VLLM_HOST_IP --port=6379 --num-cpus=64 --num-gpus=$tiles&
```

## Serve Small Models 

#### Using Single Tile

The following command serves `meta-llama/Llama-2-7b-chat-hf` on a single tile of a single node:
```bash linenums="1"
vllm serve meta-llama/Llama-2-7b-chat-hf --port 8000 --device xpu --dtype float16
```

#### Using Multiple Tiles

The following commands set up the Ray cluster and serve `meta-llama/Llama-3.3-70b-chat-hf` on 8 tiles on single node. 
```bash linenums="1"
export VLLM_HOST_IP=$(getent hosts $(hostname).hsn.cm.aurora.alcf.anl.gov | awk '{ print $1 }' | tr ' ' '\n' | sort | head -n 1)
ray --logging-level debug start --head --verbose --node-ip-address=$VLLM_HOST_IP --port=6379 --num-cpus=64 --num-gpus=8&
vllm serve meta-llama/Llama-2-7b-chat-hf --port 8000 --tensor-parallel-size 8 --device xpu --dtype float16 --trust-remote-code
```

## Serve Medium Models 

#### Using Single Node

> Following sets up ray cluster and servers `meta-llama/Llama-3.3-70B-Instruct` on 8 tiles on single node. Models with up to 70 billion parameters can usually fit within a single node, utilizing multiple tiles.

Reviewer: ```suggestion
These commands set up a Ray cluster and serves `meta-llama/Llama-3.3-70B-Instruct` on 8 tiles on single node. Models with up to 70 billion parameters can usually fit within a single node, utilizing multiple tiles.
```

```bash
export VLLM_HOST_IP=$(getent hosts $(hostname).hsn.cm.aurora.alcf.anl.gov | awk '{ print $1 }' | tr ' ' '\n' | sort | head -n 1)
ray --logging-level debug start --head --verbose --node-ip-address=$VLLM_HOST_IP --port=6379 --num-cpus=64 --num-gpus=8&
vllm serve meta-llama/Llama-3.3-70B-Instruct --port 8000 --tensor-parallel-size 8 --device xpu --dtype float16 --trust-remote-code
```

## Serve Large Models 

> ### Using multiple nodes

Reviewer: ```suggestion
### Using Multiple Nodes
```

just trying to be consistent with title case

```bash linenums="1"
module load frameworks
conda create --name vllm python=3.10 -y
conda activate vllm

module unload oneapi/eng-compiler/2024.07.30.002
module use /opt/aurora/24.180.3/spack/unified/0.8.0/install/modulefiles/oneapi/2024.07.30.002
module use /soft/preview/pe/24.347.0-RC2/modulefiles
module add oneapi/release

pip install /flare/datascience/sraskar/vllm-install/wheels/*
pip install /flare/datascience/sraskar/vllm-install/vllm-0.6.6.post2.dev28+g5dbf8545.d20250129.xpu-py3-none-any.whl
```

## Access Model Weights

Model weights for commonly used open-weight models are downloaded and available in the following directory on Aurora:
```bash
/flare/datascience/model-weights/hub
```
To ensure your workflows utilize the preloaded model weights and datasets, update the following environment variables in your session. Some models hosted on Hugging Face may be gated, requiring additional authentication. To access these gated models, you will need a [Hugging Face authentication token](https://huggingface.co/docs/hub/en/security-tokens).
```bash
export HF_HOME="/flare/datascience/model-weights/hub"
export HF_DATASETS_CACHE="/flare/datascience/model-weights/hub"

> export HF_TOKEN="YOUR_HF_TOKEN"

Reviewer 2: done
